crates/ide/src/inlay_hints.rs

```diff
@@ -193,14 +193,68 @@ fn get_bind_pat_hints(
         return None;
     }
 
-    acc.push(InlayHint {
-        range: pat.syntax().text_range(),
-        kind: InlayKind::TypeHint,
-        label: ty.display_truncated(sema.db, config.max_length).to_string().into(),
-    });
+    let db = sema.db;
+    if let Some(hint) = hint_iterator(db, config, &ty, pat.clone()) {
+        acc.push(hint);
+    } else {
+        acc.push(InlayHint {
+            range: pat.syntax().text_range(),
+            kind: InlayKind::TypeHint,
+            label: ty.display_truncated(db, config.max_length).to_string().into(),
+        });
+    }
+
     Some(())
 }
 
+/// Checks if the type is an Iterator from std::iter and replaces its hint with an `impl Iterator<Item = Ty>`.
+fn hint_iterator(
+    db: &RootDatabase,
+    config: &InlayHintsConfig,
+    ty: &Type,
+    pat: ast::IdentPat,
+) -> Option<InlayHint> {
+    let strukt = ty.as_adt()?;
+    let krate = strukt.krate(db)?;
+    let module = strukt.module(db);
+    if krate.declaration_name(db).as_deref() != Some("core") {
+        return None;
+    }
+    let module = module
+        .path_to_root(db)
+        .into_iter()
+        .rev()
+        .find(|module| module.name(db) == Some(known::iter))?;
+    let iter_trait = module.scope(db, None).into_iter().find_map(|(name, def)| match def {
+        hir::ScopeDef::ModuleDef(ModuleDef::Trait(r#trait)) if name == known::Iterator => {
+            Some(r#trait)
+        }
+        _ => None,
+    })?;
```

Comment:
Rather then iterating through the scope, we should ask for `Iterator` directly. Currently, this handled by the `FamousDefs` type in assists. It is somewhat ad-hoc, long term, I think we need to make use of `rustc_diagnostic_item` attribute, but, for now, using `FamousDefs` as an API is the way to go!

Reply:
`FamousDefs` is crate private, I assume making it pub for this should be alright?

Reply:
I had to change a thing in `FamousDefs` aside from the visibility. It now also checks the crate you directly pass to it whether that is `std/core`, due to the crate the hint has already being std/core.

Reply:
And the `FIXTURE` const can't be `cfg(test)` gated anymore for some reason. The cfg just doesn't seem to work properly across crate boundaries?